File: src/multivariable_calculus_lagrange_multipliers_medium.py
"""
multivariable_calculus - lagrange_multipliers (medium)
Generated: 2026-02-22T05:37:50.028875
"""

from problem_utils import *

def generate():
    prob_type = randint(1, 5)

    if prob_type == 1:
        # f(x,y) = ax + by on circle x^2 + y^2 = r^2
        a_coeff = nonzero(-4, 4)
        b_coeff = nonzero(-4, 4)
        r_sq = choice([4, 9, 16, 25, 36, 49])
        # max = sqrt(r_sq*(a^2+b^2))
        max_val = simplify(sqrt(r_sq * (a_coeff**2 + b_coeff**2)))
        return problem(
            question=f"Find the maximum value of $f(x,y) = {latex(a_coeff*x + b_coeff*y)}$ subject to $x^2 + y^2 = {r_sq}$.",
            answer=max_val,
            difficulty=(1300, 1500),
            topic="multivariable_calculus/lagrange_multipliers",
            solution=steps(
                f"$\\nabla f = ({a_coeff}, {b_coeff})$, $\\nabla g = (2x, 2y)$",
                f"From $({a_coeff}, {b_coeff}) = \\lambda(2x, 2y)$: $x = \\frac{{{a_coeff}}}{{2\\lambda}}$, $y = \\frac{{{b_coeff}}}{{2\\lambda}}$",
                f"Substitute into constraint to get $\\lambda = \\pm\\frac{{\\sqrt{{{a_coeff**2}+{b_coeff**2}}}}}{{2\\sqrt{{{r_sq}}}}}$",
                f"Maximum value: ${latex(max_val)}$"
            )
        )

    elif prob_type == 2:
        # Minimize x^2 + y^2 on line ax + by = c
        a_coeff = nonzero(-4, 4)
        b_coeff = nonzero(-4, 4)
        c_val = nonzero(-8, 8)
        denom = a_coeff**2 + b_coeff**2
        min_val = simplify(Rational(c_val**2, denom))
        return problem(
            question=f"Find the minimum value of $f(x,y) = x^2 + y^2$ subject to ${latex(a_coeff*x + b_coeff*y)} = {c_val}$.",
            answer=min_val,
            difficulty=(1300, 1500),
            topic="multivariable_calculus/lagrange_multipliers",
            solution=steps(
                f"$2x = {a_coeff}\\lambda$, $2y = {b_coeff}\\lambda$",
                f"Substituting: $\\frac{{{a_coeff**2}+{b_coeff**2}}}{{2}}\\lambda = {c_val}$",
                f"$\\lambda = \\frac{{2 \\cdot {c_val}}}{{{denom}}}$",
                f"Minimum: $\\frac{{{c_val}^2}}{{{denom}}} = {latex(min_val)}$"
            )
        )

    elif prob_type == 3:
        # Maximize xy subject to x + y = k
        k_val = randint(4, 16)
        max_val = Rational(k_val**2, 4)
        return problem(
            question=f"Find the maximum value of $f(x,y) = xy$ subject to $x + y = {k_val}$.",
            answer=max_val,
            difficulty=(1300, 1400),
            topic="multivariable_calculus/lagrange_multipliers",
            solution=steps(
                f"$y = \\lambda$, $x = \\lambda$, so $x = y = \\frac{{{k_val}}}{{2}}$",
                f"Maximum: $xy = \\left(\\frac{{{k_val}}}{{2}}\\right)^2 = {latex(max_val)}$"
            )
        )

    elif prob_type == 4:
        # Maximize xyz on x + y + z = k (3 variables)
        k_val = randint(6, 18)
        val = Rational(k_val, 3)
        max_val = val**3
        return problem(
            question=f"Find the maximum value of $f(x,y,z) = xyz$ subject to $x+y+z = {k_val}$, $x,y,z > 0$.",
            answer=max_val,
            difficulty=(1400, 1600),
            topic="multivariable_calculus/lagrange_multipliers",
            solution=steps(
                f"$yz = \\lambda$, $xz = \\lambda$, $xy = \\lambda$ implies $x = y = z$",
                f"Constraint: $3x = {k_val}$, so $x = y = z = {latex(val)}$",
                f"Maximum: $\\left(\\frac{{{k_val}}}{{3}}\\right)^3 = {latex(max_val)}$"
            )
        )

    else:
        # Minimize x^2 + y^2 + z^2 on plane ax + by + cz = d
        a_coeff = nonzero(-2, 2)
        b_coeff = nonzero(-2, 2)
        c_coeff = nonzero(-2, 2)
        d_val = nonzero(-6, 6)
        denom = a_coeff**2 + b_coeff**2 + c_coeff**2
        lambda_val = Rational(2*d_val, denom)
        x_crit = simplify(lambda_val * a_coeff / 2)
        y_crit = simplify(lambda_val * b_coeff / 2)
        z_crit = simplify(lambda_val * c_coeff / 2)
        min_val = simplify(x_crit**2 + y_crit**2 + z_crit**2)
        return problem(
            question=f"Find the minimum value of $f(x,y,z) = x^2+y^2+z^2$ subject to ${latex(a_coeff*x+b_coeff*y+c_coeff*z)} = {d_val}$.",
            answer=min_val,
            difficulty=(1400, 1600),
            topic="multivariable_calculus/lagrange_multipliers",
            solution=steps(
                f"$2x = {a_coeff}\\lambda$, $2y = {b_coeff}\\lambda$, $2z = {c_coeff}\\lambda$",
                f"Substituting into constraint: $\\lambda = {latex(lambda_val)}$",
                f"Critical point: $({latex(x_crit)}, {latex(y_crit)}, {latex(z_crit)})$",
                f"Minimum: ${latex(min_val)}$"
            )
        )

emit(generate())

File: src/multivariable_calculus_lagrange_multipliers_medium.jl
# multivariable_calculus - lagrange_multipliers (medium)
# Generated: 2026-03-08T21:00:19.235843

include(joinpath(@__DIR__, "..", "julia", "src", "ProblemUtils.jl"))
using .ProblemUtils

@script x y z begin
    set_topic!("multivariable_calculus/lagrange_multipliers")
    
    prob_type = choice(1:4)
    
    if prob_type == 1
        # Optimize f(x,y) = x² + y² subject to x + y = k (1200-1400)
        k = randint(4, 12)
        ans_x = k // 2
        ans_y = k // 2
        ans_val = 2 * (k // 2)^2
        
        problem(
            question="Use Lagrange multipliers to find the minimum value of \$f(x,y) = x^2 + y^2\$ subject to the constraint \$x + y = $k\$. Enter the minimum value.",
            answer=ans_val,
            difficulty=(1200, 1400),
            solution=steps(
                "Set up Lagrange equations: \$\\nabla f = \\lambda \\nabla g\$ where \$g(x,y) = x + y - $k\$",
                sol("Equations", "2x = \\lambda, \\quad 2y = \\lambda, \\quad x + y = $k"),
                "From first two equations: \$2x = 2y\$, so \$x = y\$",
                "Substitute into constraint: \$2x = $k\$, giving \$x = y = $(k//2)\$",
                sol("Minimum value", "f($(k//2), $(k//2)) = $(ans_val)")
            ),
            time=180
        )
        
    elseif prob_type == 2
        # Optimize f(x,y) = ax² + by² subject to cx + dy = k (1300-1500)
        a = randint(1, 4)
        b = randint(1, 4)
        c = randint(2, 5)
        d = randint(2, 5)
        k = randint(6, 15)
        
        # From gradient equations: 2ax = λc, 2by = λd
        # So x = λc/(2a), y = λd/(2b)
        # Constraint: c(λc/(2a)) + d(λd/(2b)) = k
        # λ(c²/(2a) + d²/(2b)) = k
        lambda_val = k / (c^2/(2*a) + d^2/(2*b))
        ans_x = lambda_val * c / (2*a)
        ans_y = lambda_val * d / (2*b)
        ans_val = a * ans_x^2 + b * ans_y^2
        
        problem(
            question="Use Lagrange multipliers to find the minimum value of \$f(x,y) = $(a)x^2 + $(b)y^2\$ subject to \$$(c)x + $(d)y = $k\$. Round to two decimal places.",
            answer=round(ans_val, digits=2),
            difficulty=(1300, 1500),
            solution=steps(
                "Set up: \$\\nabla f = \\lambda \\nabla g\$ where \$g(x,y) = $(c)x + $(d)y - $k\$",
                sol("System", "$(2*a)x = $(c)\\lambda, \\quad $(2*b)y = $(d)\\lambda, \\quad $(c)x + $(d)y = $k"),
                "Express x and y in terms of λ: \$x = \\frac{$(c)\\lambda}{$(2*a)}\$, \$y = \\frac{$(d)\\lambda}{$(2*b)}\$",
                "Substitute into constraint and solve for λ, then find x, y",
                sol("Minimum value", "\\approx $(round(ans_val, digits=2))")
            ),
            time=240,
            calculator="scientific"
        )
        
    elseif prob_type == 3
        # Optimize f(x,y,z) = x + y + z subject to x² + y² + z² = r² (1400-1600)
        r = randint(5, 12)
        ans_coord = r / sqrt(3)
        ans_val = 3 * ans_coord
        
        problem(
            question="Use Lagrange multipliers to find the maximum value of \$f(x,y,z) = x + y + z\$ subject to \$x^2 + y^2 + z^2 = $(r^2)\$. Round to two decimal places.",
            answer=round(ans_val, digits=2),
            difficulty=(1400, 1600),
            solution=steps(
                "Set up: \$\\nabla f = \\lambda \\nabla g\$ where \$g(x,y,z) = x^2 + y^2 + z^2 - $(r^2)\$",
                sol("System", "1 = 2\\lambda x, \\quad 1 = 2\\lambda y, \\quad 1 = 2\\lambda z, \\quad x^2 + y^2 + z^2 = $(r^2)"),
                "From first three equations: \$x = y = z = \\frac{1}{2\\lambda}\$",
                "Substitute into constraint: \$3\\left(\\frac{1}{2\\lambda}\\right)^2 = $(r^2)\$",
                "Solve: \$x = y = z = \\frac{$r}{\\sqrt{3}}\$ (for maximum)",
                sol("Maximum value", "$(round(ans_val, digits=2))")
            ),
            time=240,
            calculator="scientific"
        )
        
    else
        # Optimize f(x,y,z) = xyz subject to x + y + z = s (1500-1700)
        s = randint(12, 24)
        ans_coord = s // 3
        ans_val = (s // 3)^3
        
        problem(
            question="Use Lagrange multipliers to find the maximum value of \$f(x,y,z) = xyz\$ subject to \$x + y + z = $s\$ where \$x, y, z > 0\$. Enter the exact value.",
            answer=ans_val,
            difficulty=(1500, 1700),
            solution=steps(
                "Set up: \$\\nabla f = \\lambda \\nabla g\$ where \$g(x,y,z) = x + y + z - $s\$",
                sol("System", "yz = \\lambda, \\quad xz = \\lambda, \\quad xy = \\lambda, \\quad x + y + z = $s"),
                "From \$yz = xz\$: either \$z = 0\$ or \$y = x\$. Since \$z > 0\$, we have \$y = x\$",
                "Similarly from \$yz = xy\$: \$z = x\$. Thus \$x = y = z\$",
                "Substitute: \$3x = $s\$, so \$x = y = z = $(s//3)\$",
                sol("Maximum value", "$(ans_val)")
            ),
            time=240
        )
    end
end